File: java8/src/main/java/my/rinat/kata/collections/donut/Delivery.java
package my.rinat.kata.collections.donut;

import java.time.LocalDate;

import org.eclipse.collections.api.list.ImmutableList;

public class Delivery {
    private final Order order;
    private final ImmutableList<Donut> donuts;

    public Delivery(Order order, ImmutableList<Donut> donuts) {
        this.donuts = donuts;
        this.order = order;
        this.order.getCustomer().addDelivery(this);
    }

    public Customer getCustomer() {
        return this.order.getCustomer();
    }

    public Order getOrder() {
        return this.order;
    }

    public ImmutableList<Donut> getDonuts() {
        return this.donuts;
    }

    public LocalDate getDate() {
        return this.order.getDate();
    }

    public boolean deliveredOn(LocalDate date) {
        return this.getDate().equals(date);
    }

    public int getTotalDonuts() {
        return this.getDonuts().size();
    }

    public double getTotalPrice() {
        return this.getDonuts().sumOfDouble(Donut::getPrice);
    }

    @Override
    public String toString() {
        return "Delivery(" +
                "order=" + this.order +
                ", donuts=" + this.donuts.toBag().toStringOfItemToCount() +
                ')';
    }
}

File: kotlin/kotlin-eclipse-collections-katas/src/main/kotlin/my/rinat/kata/donutkatakotlin/Delivery.kt
package my.rinat.kata.donutkatakotlin

import org.eclipse.collections.api.list.ImmutableList
import java.time.LocalDate

class Delivery(private val order: Order, val donuts: ImmutableList<Donut>) {
    init {
        this.order.customer.addDelivery(this)
    }

    val customer: Customer
        get() = this.order.customer

    fun getOrder(): Order {
        return this.order
    }

    val date: LocalDate
        get() = this.order.date

    fun deliveredOn(date: LocalDate): Boolean {
        return this.date == date
    }

    val totalDonuts: Int
        get() = this.donuts.size()

    val totalPrice: Double
        get() = this.donuts.sumOfDouble { it.price }

    override fun toString(): String {
        return "Delivery(" +
                "order=" + order +
                ", donuts=" + donuts.toBag().toStringOfItemToCount() +
                ')'
    }
}
